Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm

```diff
@@ -65,6 +69,122 @@
     return nullString();
 }
 
+void WebExtensionController::getDataRecords(OptionSet<WebExtensionDataType> types, CompletionHandler<void(Vector<Ref<WebExtensionDataRecord>>)>&& completionHandler)
+{
+    if (!m_configuration->storageIsPersistent() || types.isEmpty()) {
+        completionHandler({ });
+        return;
+    }
+
+    auto recordHolder = RecordHolder::create();
+    auto aggregator = MainRunLoopCallbackAggregator::create([recordHolder, completionHandler = WTFMove(completionHandler)]() mutable {
+        Vector<Ref<WebExtensionDataRecord>> records;
+        for (auto entry : recordHolder->recordsMap)
+            records.append(entry.value);
+
+        completionHandler(records);
+    });
+
+    auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
+    for (auto& uniqueIdentifier : uniqueIdentifiers) {
+        auto storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
```

Comment:
This can be a private helper method that is used here (and elsewhere in this file).